### ISSUE TITLE
Support Socket With gorilla/mux package

### ISSUE THREAD

**Issue by Code-Crash** [googollee/go-socket.io#339]
Hi Team,

Is there any way to connect with gorilla/mux and attaching the router with it and socket path.

I tried using mux but it's not working, please find the code below

```go
func _Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		allowHeaders := "Accept, Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization"

		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Access-Control-Allow-Origin", "http://localhost:8080")
		w.Header().Set("Access-Control-Allow-Methods", "POST, PUT, PATCH, GET, DELETE")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")
		w.Header().Set("Access-Control-Allow-Credentials", "true")
		w.Header().Set("Access-Control-Allow-Headers", allowHeaders)

		next.ServeHTTP(w, r)
	})
}

func main() {
	// Socket Setup
	server, err := socketio.NewServer(nil)
	if err != nil {
		log.Fatal(err)
	}
	socket.Initiate(server) // This code will handle the socket events, so not adding it
	go server.Serve()
	defer server.Close()
	// API Router Setup
	r := mux.NewRouter()
	r.Handle("/socket.io/", _Middleware(server))
	log.Println("Serving at localhost:8080...")
	log.Fatal(http.ListenAndServe(":8080", r))
}
```

Frontend: 

```html
<body>
    <ul id="messages"></ul>
    <form action="">
        <input id="m" autocomplete="off" /><button>Send</button>
    </form>
    <script src="https://cdn.socket.io/socket.io-1.2.0.js"></script>
    <script src="https://code.jquery.com/jquery-1.11.1.js"></script>
    <script>
        var socket = io("http://localhost:8080/socket.io", { transports: ['websocket'] });
        // socket.emit('msg', 'hello');
        var s2 = io("http://localhost:8080/socket.io/chat", { transports: ['websocket'] });
        socket.on('reply', function (msg) {
            $('#messages').append($('<li>').text(msg));
        });
        $('form').submit(function () {
            s2.emit('msg', $('#m').val(), function (data) {
                $('#messages').append($('<li>').text('ACK CALLBACK: ' + data));
            });
            socket.emit('notice', $('#m').val());
            $('#m').val('');
            return false;
        });
    </script>
</body>
```

When I'm running my frontend, I'm getting 
`http: superfluous response.WriteHeader call from github.com/googollee/go-engine%2eio.(*Server).ServeHTTP`

**Update:**

I'm able to run the project on same server, so from different server CORS is not working.

Thanks

**Comment by exapsy** [comment]
This error is an `go-engine.io` issue. The reason you can this is because:

1. HTTP upgrades to websocket in [go-engine.io / transport / websocket / transport.go #76](https://github.com/googollee/go-engine.io/blob/master/transport/websocket/transport.go#L76)
2. It upgrades with a `CheckOrigin` of `nil` (no check origin is actually set`
3. gorilla/websocket `upgrader.Upgrade(..)` sees that there's no `CheckOrigin` function, so it sets its default one.
4. The default one checks if server's host is same as frontends (eg. localhost vs localhost:81 are not same)
5. Sends an error
6. An error is send when already WriteHeader() has been called, which means that a status code has been already sent and sending another status code results in a `superfluous call`

**Solution?**
Agh. I dunno man. Not yet. Temporary replacing the code in `vendor/googollee/go-engine.io/transport/websocket/transport.go` solves the issue locally, if you replace it as such:
```go
upgrader := websocket.Upgrader{
		ReadBufferSize:  t.ReadBufferSize,
		WriteBufferSize: t.WriteBufferSize,
		CheckOrigin:     t.CheckOrigin,
	}
```
with this
```go
upgrader := websocket.Upgrader{
		ReadBufferSize:  t.ReadBufferSize,
		WriteBufferSize: t.WriteBufferSize,
		CheckOrigin:     func(r *http.Request) bool {
                        return true
                },
	}
```

But I'm still investigating to see what's really the issue here, and just documenting this stuff in here.

**Comment by Code-Crash** [comment]
Thanks @exapsy for response.

Sure, I will try this, looking at the code, it will surely work but it's hack.

What if we can create a some kind of configuration which will inform the known origins and it will return true based on that?
Again, I'm just throwing some random ideas as I'm noob in go as of now.

Thanks by the way. :) 

Please let me know, if anyone finds a proper solutions.

**Comment by exapsy** [comment]
@Code-Crash I'll try to find time to make a PR soon, so we can insert a set of allowed `origins` and then test them in a custom `CheckOrigin`.

But for the time being, you can follow the following guide I've written to use your own fork of the repository instead of the original one without making big changes.
[Debugging Golang Modules](https://github.com/exapsy/diaries/blob/master/golang%20debug%20modules/article.md)

tl;dr;
Just use this
```
go mod edit -replace github.com/googollee/go-engine.io=github.com/Code-Crash/go-engine.io
```
to point to your fork for the time being and make your own changes temporary

**Comment by Code-Crash** [comment]
Sure @exapsy , Thanks.

**Comment by erkie** [comment]
This is related to https://github.com/googollee/go-socket.io/issues/372 would love a PR to make it easier. Could also be a documentation fix to show how to override CheckOrigin. No need to use a custom fork :)
